src/coreclr/src/jit/compiler.cpp

```diff
@@ -9301,3 +9301,51 @@ void Compiler::gtChangeOperToNullCheck(GenTree* tree, BasicBlock* block)
     block->bbFlags |= BBF_HAS_NULLCHECK;
     optMethodFlags |= OMF_HAS_NULLCHECK;
 }
+
+//------------------------------------------------------------------------
+// gtTransformUnusedIndirection: change the opcode and the type of the unused indirection.
+//
+// Arguments:
+//    ind   - Indirection to transform.
+//    block - Basic block of the indirection
+//
+void Compiler::gtTransformUnusedIndirection(GenTreeIndir* ind, BasicBlock* block)
+{
+    // A nullcheck is essentially the same as an indirection with no use.
+    // The difference lies in whether a target register must be allocated.
+    // On XARCH we can generate a compare with no target register as long as the addresss
+    // is not contained.
+    // On ARM64 we can generate a load to REG_ZR in all cases.
+    // However, on ARM we must always generate a load to a register.
+    // In the case where we require a target register, it is better to use GT_IND, since
+    // GT_NULLCHECK is a non-value node and would therefore require an internal register
+    // to use as the target. That is non-optimal because it will be modeled as conflicting
+    // with the source register(s).
+    // So, to summarize:
+    // - On ARM64, always use GT_NULLCHECK for a dead indirection.
+    // - On ARM, always use GT_IND.
+    // - On XARCH, use GT_IND if we have a contained address, and GT_NULLCHECK otherwise.
+    // In all cases, change the type to TYP_INT.
+    //
+    assert(ind->OperIs(GT_NULLCHECK, GT_IND));
```

Comment:
I think that the logic for when to generate `GT_IND` vs. `GT_NULLCHECK` shoud remain in `Lowering` - see `Lowering::LowerIndir()`. Also, I found that if I changed dead indirections to `GT_NULLCHECK` too early in the compile, CSE would miss out on opportunities to CSE dead indirections in loops.

Reply:
I need to call this function from liveness that’s running during lower, that’s why I placed on Compiler. If I kept it in Lower, I would have to pass an instance of Lower to liveness, which would be odd.

Reply:
@CarolEidt it appeared that these unused `IND` could appear after lowering so we can't rely on `LowerIndir` to replace `NULLCHECK` with `IND` for arm32.

Reply:
I see. That said, I still think that we should 1) avoid generating `GT_NULLCHECK` when not in `LIR` form, and 2) leave the method on `Lowering` that determines whether to transform to `GT_NULLCHECK` as it is target-dependent. It can be static, and then this method can do the actual transformation, since it updates block and `Compiler` state (that said, it seems to me that if we care about whether the method or block has a nullcheck, that should be set for the `GT_IND` form of nullcheck as well). As much as possible, I'd like to keep the target-dependent code in the backend files.

Reply:
> Also, I found that if I changed dead indirections to GT_NULLCHECK too early in the compile, CSE would miss out on opportunities to CSE dead indirections in loops.


we have optimizations in earlyProp that prefer `NULLCHECK` over `IND`, see `OMF_HAS_NULLCHECK`. So if we stop generating `NULLCHECK` before lowering we would get both regressions and improvements. 

Also, the importer generates null checks in some cases instead of `GT_IND` and `GT_FIELD`.

I agree that it is error-prone and confusing, would be nice to find a better way to handle this.

Reply:
1) We intentionally generate `GT_NULLCHECK` before LIR in many places because both earlyProp (`gtFoldNullCheck`) ad assertion propagation have optimizations to eliminate redundant null checks.
2) I can make the method static on Lower and pass Compiler and block instances to it. Would that address your concern?

Reply:
Currently the nullcheck flags on the compiler instance and basic block are only used by earlyProp. Still, I'd like to keep them up-to-date.

Reply:
@erozenfeld - that would be fine, but it would also be fine to have a static method on `Lowering` that simply determined which option to use.

Reply:
@CarolEidt I addressed your feedback. PTAL.